DVDInterface: Translate and align offsets for timing purposes

JosJuice opened the pull request:
This is a fairly large timing change for decrypted reads and a smaller timing change for non-decrypted reads.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/1900)

<!-- Reviewable:end -->

magumagu commented on Source/Core/Core/HW/DVDInterface.cpp:
```diff
@@ -1291,7 +1343,7 @@ u64 SimulateDiscReadTime(u64 offset, u32 length)
 	}
 
-	g_last_read_offset = (offset + length - 2048) & ~2047;
+	g_last_read_offset = end_offset;
```
What's the justification for changing this computation?

JosJuice replied:
Sector alignment is now done at the top of this function, so I thought it would be unnecessary to do it again.

magumagu replied:
Sure, but you're still skipping the "- 2048" part.

JosJuice replied:
What's the point of the "- 2048"? Isn't g_last_read_offset supposed to be the last offset that has been read, not an offset before it?

magumagu replied:
I did my testing a long time ago, but IIRC the idea was to make sure that very small sequential reads get covered by the "buffered" codepath.

JosJuice replied:
Okay, so the current code essentially has the same problem as the `if (decrypt) ...` code. I think modifying the code for buffered reads would be better than keeping the "- 2048" here.

To clarify, does the hardware only let you use the buffer if you're reading within 2048 bytes back, or is it possible to read from the buffer further back than that too?

magumagu replied:
IIRC, the buffer essentially only works forward; I don't recall exactly how far back it lets you go.  If you want to experiment, I can dig up my old hwtest.

JosJuice replied:
What I was thinking of was making it possible to go back to data that has been read, as long as it hasn't already been overwritten by new data (which would take 0,5 s on 2 MiB/s). It would require us to do some additional bookkeeping, though, so that it isn't possible to go back to data from before the read started.

magumagu replied:
I'm pretty sure that doesn't match what the hardware does., but feel free to prove me wrong.

If you want to experiment, my old DVD testing code I used to discover the buffering is here: https://github.com/magumagu/hwtests/blob/dvdhack/gxtest/source/main.cpp#L3001 .  Should be straightforward to compile/run/modify.  You can ping me on IRC if you have questions, or if you need me to run it for some reason.

There's one important question here: does the DVD drive actually perform decryption, or does IOS do it?  If IOS does it, there's probably another layer of caching we need to account for to compute correct timings.

JosJuice replied:
Okay. I will stick to something "- 2048"-like then.

IOS is doing the decryption. It would be nice to be able to put Dolphin's decryption in WII_IPC_HLE_Device_DI, but it won't work with the way DiscIO is set up to do it.

JosJuice replied:
Actually, screw it. Keeping the "- 2048" where it was is probably the best. I'll add something separate for buffering decrypted reads later.

magumagu replied:
No matter where we actually perform the decryption, we should still be able to account for the relevant timing differences: the IPCHLE code still knows that decryption is happening even if it doesn't happen there.

magumagu commented on Source/Core/Core/HW/DVDInterface.cpp:
```diff
@@ -1215,8 +1216,60 @@ DVDCommandResult ExecuteCommand(u32 command_0, u32 command_1, u32 command_2,
+	// Offsets on disc layer 2 are replaced with offsets on disc layer 1
+	// to aid in calculations that need to use physical positions.
+	offset %= WII_DISC_LAYER_SIZE;
+	end_offset %= WII_DISC_LAYER_SIZE;
```
Is WII_DISC_LAYER_SIZE the size of an encrypted layer or an unencrypted layer?  Either way, you're using it wrong.

JosJuice replied:
Unencrypted.

tueidj replied:
From the comment ("Offsets on disc layer 2 ...") and my memories of implementing the speed estimation in cleanrip, this behavior is wrong because dual layer wii discs use OTP (opposite track path) meaning the second layer starts on the outside of the disc rather than the inside.

JosJuice replied:
Ah, thanks. I will fix it.

magumagu commented on Source/Core/Core/HW/DVDInterface.cpp:
```diff
@@ -1215,8 +1216,60 @@ DVDCommandResult ExecuteCommand(u32 command_0, u32 command_1, u32 command_2,
+			end_offset = offset;
+		else
+			end_offset = (end_offset + 0x7BFF) / 0x7C00 * 0x8000 + data_start;
+	}
```
Are you sure the math actually works like this for "buffered" reads?  This might mess up timings for small sequential reads.

JosJuice replied:
I assume you're talking about the case where sequential reads access the same block? I assumed that it would be placed in the cache on the first read and then re-read from there, but now that I'm taking a closer look, the cache doesn't seem to work like that. Thanks for catching it.

JosJuice commented:
I've added a fix for buffered decrypted reads. Not sure how good it is.

delroth commented:
@dolphin-emu-bot rebuild (sorry, was doing some buildbot maintenance at the same time)

Buddybenj commented:
Ping?

skidau commented:
How did you test this? How did you tell that this is more accurate?

JosJuice commented:
I don't have a good way to test this, really... :/ I guess it would be possible to compare the loading times in some Wii game, but if we want something less approximate, writing custom homebrew software would probably be needed. CleanRip won't help with this, because it doesn't decrypt and it reads everything using big, aligned requests. But while I don't know how accurate things are with this PR, I know that the old way of completely ignoring decryption can't be right. It would require the disc drive to magically read faster whenever IOS wants to decrypt data.

shuffle2 commented:
It's hard to say if this is more accurate or not. My guess is that overall it's not really more accurate, since the proper simulation would be having our IOS HLE simulate IOS' dvd read strategy when files are accessed.

So, I'm not really in favor of the commit. However, if it is going to be committed, please make all the magic sizes be constant variables so that they have names.

JosJuice commented:
The constants from VolumeWiiCrypted are now used, along with a SECTOR_SIZE constant.

JosJuice commented:
A hardware test by @mmastrac showed that disc drive reads are aligned to 32 KiB, the size of an ECC block, which makes sense. Because of that, I've changed this PR from aligning to 2 KiB to aligning to 32 KiB. Also, I removed the `- 2048` because it didn't exist in the test results as far as I understood them. There is still code that ensures that small sequential reads will be covered by the buffered code path.

JosJuice commented:
PR #4829 makes half of this PR obsolete. The other half needs to be more or less rewritten. I'll open a new PR for that later.
